phoenix-core/src/main/java/org/apache/phoenix/schema/transform/TransformMaintainer.java

```diff
@@ -0,0 +1,459 @@
+    /*
+     * The first part of the pair is column family name
+     * and second part is the column name. The reason we need to track this state is because for certain storage schemes
+     * like ImmutableStorageScheme#SINGLE_CELL_ARRAY_WITH_OFFSETS, the column for which we need to generate an new
+     * table put/delete is different from the old columns in the phoenix schema.
+     */
+    private Set<Pair<String, String>> newTableColumnsInfo;
+    /*
+     * Map of covered columns where a key is column reference for a column in the data table
+     * and value is column reference for corresponding column in the new table.
+     */
+    private Map<ColumnReference, ColumnReference> coveredColumnsMap;
+
+    private String logicalNewTableName;
+
+    public static TransformMaintainer create(PTable oldTable, PTable newTable, PhoenixConnection connection) {
+        if (oldTable.getType() == PTableType.INDEX) {
+            throw new IllegalArgumentException();
+        }
+        TransformMaintainer maintainer = new TransformMaintainer(oldTable, newTable, connection);
+        return maintainer;
+    }
+
+    private TransformMaintainer(RowKeySchema oldRowKeySchema, boolean isOldTableSalted) {
+        super(oldRowKeySchema, isOldTableSalted);
+        this.oldTableRowKeySchema = oldRowKeySchema;
+        this.isOldTableSalted = isOldTableSalted;
+    }
+
+    private TransformMaintainer(final PTable oldTable, final PTable newTable, PhoenixConnection connection) {
+        this(oldTable.getRowKeySchema(), oldTable.getBucketNum() != null);
+        this.newTableRowKeyOrderOptimizable = newTable.rowKeyOrderOptimizable();
+        this.isMultiTenant = oldTable.isMultiTenant();
+
+        this.newTableEncodingScheme = newTable.getEncodingScheme() == null ? PTable.QualifierEncodingScheme.NON_ENCODED_QUALIFIERS : newTable.getEncodingScheme();
+        this.newTableImmutableStorageScheme = newTable.getImmutableStorageScheme() == null ? PTable.ImmutableStorageScheme.ONE_CELL_PER_COLUMN : newTable.getImmutableStorageScheme();
+        this.oldTableEncodingScheme = oldTable.getEncodingScheme() == null ? PTable.QualifierEncodingScheme.NON_ENCODED_QUALIFIERS : oldTable.getEncodingScheme();
+        this.oldTableImmutableStorageScheme = oldTable.getImmutableStorageScheme() == null ? PTable.ImmutableStorageScheme.ONE_CELL_PER_COLUMN : oldTable.getImmutableStorageScheme();
+
+        this.newTableName = newTable.getPhysicalName().getBytes();
+        boolean newTableWALDisabled = newTable.isWALDisabled();
+        int nNewTableColumns = newTable.getColumns().size();
+        int nNewTablePKColumns = newTable.getPKColumns().size();
+
+        List<PColumn> oldTablePKColumns = oldTable.getPKColumns();
+
+        this.newTableColumnCount = oldTablePKColumns.size();
+
+        this.newTableColumnTypes = Lists.<PDataType>newArrayListWithExpectedSize(nNewTablePKColumns);
+        this.newTableExpressions = Lists.newArrayListWithExpectedSize(nNewTableColumns);
```

Comment:
nit: remove type arg from `newTableColumnTypes` as well (similar to `newTableExpressions`)?